### ISSUE TITLE
Support sync flatten for an array of generic lazy futures

### ISSUE THREAD

**Issue by wlisac** [vapor/async#82]
Currently it's only possible to `syncFlatten` an array of `LazyFuture`s where the result type is `Void`:

```swift
/// A closure that returns a future.
public typealias LazyFuture<T> = () throws -> Future<T>

extension Collection where Element == LazyFuture<Void> {
    public func syncFlatten(on worker: Worker) -> Future<Void> { … }
}
```

I'd like to be able to `syncFlatten` for any result type, similar to `flatten`, but serially.

Below is a first pass at a working implementation:

```swift
/// A lazy future type.
/// Concretely implemented by `LazyFuture<Expectation>`
public protocol LazyFutureType {
    /// See `FutureType`.
    associatedtype Expectation
    
    /// Returns a new future.
    func makeFuture() throws -> Future<Expectation>
}

public struct LazyFuture<Expectation>: LazyFutureType {
    private let body: () throws -> Future<Expectation>
    
    /// Creates a lazy future that wraps the closure in its `makeFuture()` method.
    public init(_ body: @escaping () throws -> Future<Expectation>) {
        self.body = body
    }
    
    public func makeFuture() throws -> Future<Expectation> {
        return try body()
    }
}

extension Collection where Element: LazyFutureType {
    /// Flattens an array of lazy futures into a future with an array of results.
    /// - note: each subsequent future will wait for the previous to complete before starting.
    public func syncFlatten(on worker: Worker) -> Future<[Element.Expectation]> {
        let promise = worker.eventLoop.newPromise([Element.Expectation].self)
        
        // allocate results array
        var results = [Element.Expectation]()
        
        var iterator = makeIterator()
        func handle(_ lazy: Element) {
            do {
                try lazy.makeFuture().do { res in
                    results.append(res)
                    if let next = iterator.next() {
                        handle(next)
                    } else {
                        promise.succeed(result: results)
                    }
                    }.catch { error in
                        promise.fail(error: error)
                }
            } catch {
                promise.fail(error: error)
            }
        }
        
        if let first = iterator.next() {
            handle(first)
        } else {
            promise.succeed(result: results)
        }
        
        return promise.futureResult
    }
}
```

I'd be happy to open a more comprehensive PR if this is something that others are interested in, too.

One issue is that this is a breaking change, since the new `LazyFuture` `struct` conflicts with the existing `LazyFuture` ` typealias`.

Any feedback is appreciated. Cheers!

**Comment by wlisac** [comment]
As an aside, this implementation is heavily based on the current `syncFlatten` implementation for `LazyFuture<Void>` collections.

As an alternative implementation, it may be worth exploring using `reduce` on the collection to reduce the `LazyFuture<T>`s into a single `Future<[T]>`.

**Comment by vzsg** [comment]
You don't need to make any breaking changes, nor introduce a new struct just to make this functionality possible. Previously on Discord, I've pitched my implementation of syncFlatten for the existing LazyFuture typealias:

```swift
extension Sequence {
    func syncFlatten<T>(on worker: Worker) -> Future<[T]> where Self.Element == LazyFuture<T> {
        let promise = worker.eventLoop.newPromise([T].self)
        var iterator = makeIterator()
        var result: [T] = []

        func handle(_ future: LazyFuture<T>) {
            do {
                try future().do { res in
                    result.append(res)

                    guard let next = iterator.next() else {
                        promise.succeed(result: result)
                        return
                    }

                    handle(next)
                    }.catch { error in
                        promise.fail(error: error)
                }
            } catch {
                promise.fail(error: error)
            }
        }

        if let first = iterator.next() {
            handle(first)
        } else {
            promise.succeed(result: result)
        }

        return promise.futureResult
    }
}
```

But it has been forgotten since. It's nice to have the topic in plain sight again.

**Comment by wlisac** [comment]
Ah, of course. I’m not sure how I missed this approach the first time around. 

Great update. 

Given the additive nature of this change, I think we can move towards PR.

Thanks for sharing!

**Comment by wlisac** [comment]
One minor downside to having an extension on `Sequence` or `Collection` that contains a method with only a same type constraint is that it ends up "polluting" the autocomplete for unrelated types.

```swift
extension Sequence {
    func syncFlatten<T>(on worker: Worker) -> Future<[T]> where Self.Element == LazyFuture<T> { ... }
}
```

That signature will result in an invalid autocomplete suggestion for an array of strings, for example:

<img width="464" alt="screen shot 2018-11-18 at 9 34 33 am" src="https://user-images.githubusercontent.com/2246764/48676043-d8dcf500-eb15-11e8-8c4a-aa6af75af3e7.png">

Using a protocol in the extension where clause is one workaround for this code completion issue. So something like this won't "pollute" autocomplete:

```swift
extension Collection where Element: LazyFutureType { … }
```

I believe the related autocomplete bug has been captured in [SR-7046](https://bugs.swift.org/browse/SR-7046).

It still seems to me like the non-breaking change is a better solution for now, with possibly moving to a `LazyFutureType` `protocol` and `LazyFuture` `struct` in a major update.
